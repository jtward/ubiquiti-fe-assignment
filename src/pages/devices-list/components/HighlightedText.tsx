import { PropsWithChildren } from 'react';
import { Fragment } from 'react/jsx-runtime';

interface HighlightedTextProps {
  text: string;
  query: string;
  MatchComponent: React.ComponentType<PropsWithChildren<{ key?: string }>>;
}

export const HighlightedText: React.FC<HighlightedTextProps> = ({ text, query, MatchComponent }) => {
  if (!query || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the query and create a case-insensitive regex
  // const escapedQuery = query.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const regex = new RegExp(`(${query})`, 'gi');

  // Split the text into parts based on the regex matches
  const parts = text.split(regex);

  return (
    <>
      {parts.map((part, index) => {
        const isMatch = regex.test(part);
        const Component = isMatch ? MatchComponent : Fragment;
        // eslint-disable-next-line react-x/no-array-index-key
        return <Component key={`${index}`}>{part}</Component>;
      })}
    </>
  );
};