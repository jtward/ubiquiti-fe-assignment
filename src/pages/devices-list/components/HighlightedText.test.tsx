import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { HighlightedText } from './HighlightedText';
import { PropsWithChildren } from 'react';

const MatchComponent: React.FC<PropsWithChildren> = ({ children }) => {
  return (<span>{ children }</span>);
};

describe('HighlightedText', () => {
  it('renders with no match', () => {
    const testId = 'A';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="foo"
          query="bar"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId('A')).toMatchSnapshot();
  });

  it('renders with full match', () => {
    const testId = 'B';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="foo"
          query="foo"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId(testId)).toMatchSnapshot();
  });

  it('renders with partial match', () => {
    const testId = 'C';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="a foo b"
          query="foo"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId(testId)).toMatchSnapshot();
  });

  it('renders with partial match in word', () => {
    const testId = 'C';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="afoob"
          query="foo"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId(testId)).toMatchSnapshot();
  });

  it('renders with multiple matches', () => {
    const testId = 'D';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="a foo b foo"
          query="foo"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId(testId)).toMatchSnapshot();
  });

  it('renders with adjacent matches', () => {
    const testId = 'E';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="a foofoo"
          query="foo"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId(testId)).toMatchSnapshot();
  });

  it('renders with matches containing regex characters', () => {
    const testId = 'F';
    render(
      <span data-testId={testId}>
        <HighlightedText
          text="a f$o\[o"
          query="f$o\[o"
          MatchComponent={MatchComponent}
        />
      </span>
    );
    expect(screen.getAllByTestId(testId)).toMatchSnapshot();
  });
});
