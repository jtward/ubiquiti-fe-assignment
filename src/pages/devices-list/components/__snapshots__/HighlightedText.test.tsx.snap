// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HighlightedText > renders with adjacent matches 1`] = `
[
  <span
    data-testid="E"
  >
    a 
    <span>
      foo
    </span>
    <span>
      foo
    </span>
  </span>,
]
`;

exports[`HighlightedText > renders with full match 1`] = `
[
  <span
    data-testid="B"
  >
    <span>
      foo
    </span>
  </span>,
]
`;

exports[`HighlightedText > renders with matches containing regex characters 1`] = `
[
  <span
    data-testid="F"
  >
    a 
    <span>
      f$o\\[o
    </span>
  </span>,
]
`;

exports[`HighlightedText > renders with multiple matches 1`] = `
[
  <span
    data-testid="D"
  >
    a 
    <span>
      foo
    </span>
     b 
    <span>
      foo
    </span>
  </span>,
]
`;

exports[`HighlightedText > renders with no match 1`] = `
[
  <span
    data-testid="A"
  >
    foo
  </span>,
]
`;

exports[`HighlightedText > renders with partial match 1`] = `
[
  <span
    data-testid="C"
  >
    a 
    <span>
      foo
    </span>
     b
  </span>,
]
`;

exports[`HighlightedText > renders with partial match in word 1`] = `
[
  <span
    data-testid="C"
  >
    a
    <span>
      foo
    </span>
    b
  </span>,
]
`;
